### ISSUE TITLE
graphs must be unrolled to align with POA (was: flipping breaks inverting edges)

### ISSUE THREAD

**Issue by ekg** [vgteam/vg#182]
This fails. The alignment gets corrupted by the node flipping process that's used to prep the graph for partial order alignment. (To be pushed soon as a test:)

```bash
vg align -s AGTCCTTGAAAGAGGGCAAAATAAACTGTTAGTAGAGCCAGGTCTGAAAACAACACTTTCTTGC inverting/m.vg >/dev/null
```

It exposed my own confusion. The inversions are being broken. This isn't quite what I'd expected (or hoped) it does.

For example. Before, input:

```json
{"node": [{"sequence": "GTATTTTTAGTA", "id": 2308}, {"sequence": "T", "id": 2309}, {"sequence": "CTAATTTTT", "id": 2310}, {"sequence": "CA", "id": 2311}, {"sequence": "GG", "id": 2312}, {"sequence": "ACGCCC", "id": 2313}, {"sequence": "C", "id": 2314}, {"sequence": "T", "id": 2315}, {"sequence": "C", "id": 173}, {"sequence": "C", "id": 2316}, {"sequence": "GAGACGGGGTTTCACCATGTT", "id": 2323}, {"sequence": "T", "id": 171}, {"sequence": "C", "id": 172}, {"sequence": "TACTAAAAATA", "id": 174}, {"sequence": "TGGC", "id": 181}, {"sequence": "A", "id": 2317}, {"sequence": "GGTCAGGCTGGTCTCGACTCC", "id": 2324}, {"sequence": "TC", "id": 170}, {"sequence": "T", "id": 175}, {"sequence": "C", "id": 176}, {"sequence": "GGTGTGG", "id": 180}, {"sequence": "AT", "id": 182}, {"sequence": "GG", "id": 183}, {"sequence": "GGGATTACAGGCGCACACC", "id": 2318}, {"sequence": "TGACCTCCTGATCTGCCCCCC", "id": 2325}], "edge": [{"from": 2309, "to": 2308}, {"from": 2310, "to": 2308}, {"from": 2308, "to": 2323}, {"from": 2310, "to": 2309}, {"from": 2311, "to": 2310}, {"from": 2312, "to": 2310}, {"from": 2313, "to": 2311}, {"from": 2313, "to": 2312}, {"from": 2314, "to": 2313}, {"from": 2315, "to": 2313}, {"from": 2316, "to": 2314}, {"from": 2316, "to": 2315}, {"from": 173, "to": 2308, "to_end": true}, {"from": 173, "to": 174}, {"from": 2317, "to": 2316}, {"from": 2323, "to": 2324}, {"from": 171, "to": 173}, {"from": 172, "to": 173}, {"from": 174, "to": 175}, {"from": 174, "to": 176}, {"from": 2316, "to": 181, "from_start": true}, {"from": 181, "to": 182}, {"from": 181, "to": 183}, {"from": 2318, "to": 2317}, {"from": 2324, "to": 2325}, {"from": 170, "to": 171}, {"from": 170, "to": 172}, {"from": 180, "to": 181}]}
```

Looks like:

![screenshot from 2016-01-15 16-14-40](https://cloud.githubusercontent.com/assets/145425/12358019/21d9c382-bba3-11e5-9e7f-6a04380371ce.png)

But [after the initial flip, things get funny](https://github.com/ekg/vg/blob/0675976923c8788bfd5e956b976326f430e99e9e/src/vg.cpp#L5416).

```json
{"node": [{"sequence": "TACTAAAAATAC", "id": 2308}, {"sequence": "A", "id": 2309}, {"sequence": "AAAAATTAG", "id": 2310}, {"sequence": "TG", "id": 2311}, {"sequence": "CC", "id": 2312}, {"sequence": "GGGCGT", "id": 2313}, {"sequence": "G", "id": 2314}, {"sequence": "A", "id": 2315}, {"sequence": "C", "id": 173}, {"sequence": "G", "id": 2316}, {"sequence": "GAGACGGGGTTTCACCATGTT", "id": 2323}, {"sequence": "T", "id": 171}, {"sequence": "C", "id": 172}, {"sequence": "TACTAAAAATA", "id": 174}, {"sequence": "TGGC", "id": 181}, {"sequence": "T", "id": 2317}, {"sequence": "GGTCAGGCTGGTCTCGACTCC", "id": 2324}, {"sequence": "TC", "id": 170}, {"sequence": "T", "id": 175}, {"sequence": "C", "id": 176}, {"sequence": "GGTGTGG", "id": 180}, {"sequence": "AT", "id": 182}, {"sequence": "GG", "id": 183}, {"sequence": "GGTGTGCGCCTGTAATCCC", "id": 2318}, {"sequence": "TGACCTCCTGATCTGCCCCCC", "id": 2325}], "edge": [{"from": 2308, "to": 2309}, {"from": 2308, "to": 2310}, {"from": 2309, "to": 2310}, {"from": 2310, "to": 2311}, {"from": 2310, "to": 2312}, {"from": 2311, "to": 2313}, {"from": 2312, "to": 2313}, {"from": 2313, "to": 2314}, {"from": 2313, "to": 2315}, {"from": 2314, "to": 2316}, {"from": 2315, "to": 2316}, {"from": 173, "to": 174}, {"from": 173, "to": 2308}, {"from": 2316, "to": 2317}, {"from": 2323, "to": 2324}, {"from": 171, "to": 173}, {"from": 172, "to": 173}, {"from": 174, "to": 175}, {"from": 174, "to": 176}, {"from": 181, "to": 182}, {"from": 181, "to": 183}, {"from": 2317, "to": 2318}, {"from": 2324, "to": 2325}, {"from": 170, "to": 171}, {"from": 170, "to": 172}, {"from": 180, "to": 181}]}
```

![screenshot from 2016-01-15 16-16-45](https://cloud.githubusercontent.com/assets/145425/12358071/62d3bb2c-bba3-11e5-987d-3defdb75e84b.png)

I honestly think I hadn't thought this all the way through.

Here's what I'd like:

Loops and inversions get unrolled into new nodes so the graph becomes partially ordered and compatible with gssw, but still contains the same sequence space as the input up to a given length of cycle expansions --- probably on the order of a read length. The translation is recorded. Then the alignment is modified so the path points at the original nodes in their correct orientation.

:tada: `VG::unroll`

**Comment by ekg** [comment]
Rendering with arrowheads helps:

![screenshot from 2016-01-15 16-49-32](https://cloud.githubusercontent.com/assets/145425/12358995/01bca862-bba8-11e5-9d1c-3c9e2ded8e75.png)

![screenshot from 2016-01-15 16-49-43](https://cloud.githubusercontent.com/assets/145425/12358999/04ca4e06-bba8-11e5-9dd2-81393133a612.png)

**Comment by ekg** [comment]
Algorithm. Unroll and un-invert the graph into a DAG to prepare for application of partial order alignment (POA) as implemented in gssw.

We start with any node and begin to write a new graph where we un-invert inverting edges and unroll cycles. We need to record how far we go from the start when doing this and break when we reach some base pair limit.

There are four edge types.

For the first two we do nothing, we just copy the structure into the new graph.

    from+ -> to+
    from- -> to-

For the inverting edges:

    from+ -> to-
    from- -> to+

We convert `from+ -> to-` to `from+ -> to+` in the new graph, step across and start inverting everything. If we are working in the other direction, and walk through a `from- -> to+` link, we change it into a `from+ -> to+` link and on the other side start inverting the graph we can reach from it.

Things get nasty if things loop, because we will be able to generate impossible flip combinations, such as by trying to unflip a flipped region of the graph. Unrolling the loops is also helpful for preparing ourselves to apply POA, so it makes sense to do here.

Loop unrolling means we will duplicate some nodes in the base graph in our new graph. To make the resulting graph useful, we have to record a translation from the new nodes into the old ones. This mapping also needs to record if the new node is a reverse complement of the first.

The traversal of the graph needs to be tracked. We want to maintain a list of steppers that traverse the graph in a kind of breadth-first search. Each bifurcation will generate a new pair of steppers from the input stepper. Rather than explode into a tree, in which each bifurcation leads to a new branch (and is inefficient to work with in POA), we want to ensure that we can merge things up again where there are non-looping bubbles. This could be accomplished by detecting the regions of the graph that will be stable under the transformation--- basically where there are no inversions and no cycles. We then aim to generate a tree based on these structures, rooted at our starting node.

What we want to do is first detect these components. We can project these components into a graph that describes their relationships with neighboring components in terms of inversions. Now, we build them into a tree that is long enough to simulate the mapping of our read to any position in the original graph (the unrolling distance). Then, we apply the appropriate reverse complementing, fix up the edges, and emit a ~tree/DAG that is the composition of all these copies of the various non-inverting, non-looping components of the graph.

**Comment by ekg** [comment]
I dug into this further yesterday with @jltsiren and @edawson.

I think I can now clarify the algorithm for unrolling the graph.

Our objective is to convert a cyclic, bidirectional graph into a unidirectional k-DAG, with some k on the order of 2x the length of the sequence we'd like to align. This allows the application of partial order alignment to the graph.

First, we find the strongly connected components of the graph using an algorithm like [Tarjan's strongly connected components algorithm](https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm). By definition, strongly connected components must have internal cycles, so we can use these to guide the unrolling process. We start at the entry points of the strongly connected components and enumerate all k-paths that go into the component (a k-path is a path whose concatenated label length is less than k). Now, for each k-path we generate a new path in which the nodes are indexed by the number of times we have traversed them previously in the path. For instance, if we had path ABBCDABC, we would write it A<sub>1</sub>B<sub>1</sub>B<sub>2</sub>C<sub>1</sub>D<sub>1</sub>A<sub>2</sub>B<sub>3</sub>C<sub>2</sub>. We now take the paths converted as such and generate a new graph from them, which is a DAG in which each uniquely indexed node becomes a new node. Each N<sub>i</sub> is, by definition, equivalent to other N<sub>i</sub>'s, in that any path crossing it has only seen N<sub>i-1</sub> or lower. This guarantees we generate an acyclic structure in which data dependencies are partially ordered. We extend this model to handle inverting edges by recording if we pass through a node in the forward or reverse direction, so the previous example might become A<sub>1</sub>B<sub>1</sub>B'<sub>1</sub>C'<sub>1</sub>D<sub>1</sub>A<sub>2</sub>B<sub>2</sub>C<sub>1</sub>, if N is the forward direction of the original node and N' is the reverse complement. In this way we handle both cyclic paths and also inversions in the unrolling process.

The result is a DAG to which we can apply partial order alignment. A mapping between the new nodes and nodes and orientations in the previous graph may be used to trivially project the mapping against the new unrolled graph back into the old one. In effect, this process enables the direct mapping of reads into cyclic, bidirectional graphs.

**Comment by jltsiren** [comment]
Here is the current sketch of the algorithm:

1. Find the strongly connected components in the graph.
1. Anything outside the strongly connected components can be copied directly into the DAG. We can also reuse the entry nodes to the strongly connected components.
1. Use backtracking search starting from each entry node of each strongly connected component, keeping track of the nodes on the path from the entry node to the current node:
  1. If the current node is outside the component, terminate this branch and return to the previous branching point.
  1. Create a new copy of the current node in the DAG and use that copy for this branch.
  1. If we have found the first cycle in the current branch (the current node is the first one occurring twice in the path), initialize path length to 1 for this branch. (We need to find a k-path starting from the last offset of the previous node.)
  1. If we have found a cycle in the current branch, increment path length by the length of the label of the current node. If path length >= k, terminate this branch and return to the previous branching point.

The algorithm creates a lot of redundant nodes, but we don't know how to merge them efficiently without creating cycles.

**Comment by ekg** [comment]
> The algorithm creates a lot of redundant nodes, but we don't know how to merge them efficiently without creating cycles.

We can use partial order alignment to merge them without generating cycles.

We'd want to use partial order to partial order alignment to minimize runtime. I don't have an implemenation of this, and I'm not sure how easy it will be to adapt something like the original [POA](http://sourceforge.net/projects/poamsa/). Also, it might not help much in terms of removing redundancy, and further it might be very expensive even if completed in `O(log(N))` steps for N sequences. It could still be useful if it were cached for the entire graph during the indexing step.

**Comment by ekg** [comment]
In examining @jltsiren's sketch above, I think there may be a problem with the restart condition:

> If we have found the first cycle in the current branch (the current node is the first one occurring twice in the path), initialize path length to 1 for this branch. (We need to find a k-path starting from the last offset of the previous node.)

It seems to me that this means we won't find cycles that are less than k away from our entry points. However, we want to capture these even if they are embedded inside the component more than k steps from the entry points. To resolve, I think it's sufficient to obtain all k-paths for all nodes in the component, and then build them into a DAG by avoiding the introduction of any edges that would create cycles--- and instead adding new nodes. If the original node is N, then we'll call the next node N', and the one after that N'' and so on.

So the updated sketch would be:

1. find the strongly connected components of the graph
2. if the component has a single node, add it to the unrolled graph
3. for each k-path of each node in the strongly connected components:
 1. make the node labels in the k-path unique, for instance ABCBC would become ABCB'C'.
 2. for each node N in the modified k-path:
   1. find the node in the unrolled graph which we are building, adding it if it does not exist
    2. connect the node to its predecessor in the path, unless doing so would introduce a cycle
    3. if connecting to the previous node in the path would introduce a cycle, add a new node N' and connect it to the previous node in the path

Working this out for the graph drawn at top left unrolls it into the graph at top right. The asterisk "*" marks indicate which nodes would get new identities in the graph due to the potential introduction of loops.

![photo_2016-01-22_11-52-44](https://cloud.githubusercontent.com/assets/145425/12510028/b9be7188-c0fe-11e5-926a-216278928a7e.jpg)

The problem with this technique is that naively checking for loops using strong component detection will cost us O(|N|+|E|) operations per path element. This doesn't need to be done for every element that's added, as unique ones will not need to be checked, but it is still painful.

**Comment by ekg** [comment]
Note that the example shows the unrolling for the entire graph, not just the strongly connected component.

**Comment by jltsiren** [comment]
Here is a simulation of my algorithm sketch in your example graph:

```
B (entry point, do not duplicate)
    C1 (not seen before)
        B1 (second visit, length = 2)
            E (outside the component, terminate)
            C2 (length = 3, terminate)
        D1 (not seen before)
            C3 (second visit, length = 2)
                B2 (length = 3, terminate)
                D2 (length = 3, terminate)
            D3 (second visit, length = 2)
                C4 (length = 3, terminate)
                D4 (length = 3, terminate)
```

In addition to the original path A -> B -> E, the acyclic graph would have two additional copies of B, four copies of C, and four copies of D. The graph would not contain the original nodes C and D.

The intuition is the following: For each entry point to each strongly connected component, we create at tree of paths starting from the entry point. To ensure that we end up with a tree, we create a new copy of each node we encounter. The paths can be of any length, but we terminate them once we get out of the component or when the current branch has continued long enough after closing its first cycle.

In the above example, B -> C1 -> D1 is a simple path, but any extension of it closes a cycle. To guarantee that the acyclic graph contains all paths of length 3 in the component, we have to find all paths of length 3 starting from the last node of the simple path (D1) and continuing with a successor that closes a cycle. Hence the length of the path is initialized as 1 + length(C3) = 2 and 1 + length(D3) in the successor nodes C3 and D3.

In the same way, B -> C1 is a simple path. Continuing to B1 closes a cycle, so we need to find all paths of length 3 starting with C1 -> B1.

**Comment by ekg** [comment]
Thanks @jltsiren, that clarifies things. I'm going to implement this today.

**Comment by ekg** [comment]
I have an initial implementation of this. Here is the result for the strongly connected component in the graph example I gave earlier.

The graph:

![screenshot from 2016-01-26 12-40-21](https://cloud.githubusercontent.com/assets/145425/12579695/2337d176-c42a-11e5-935a-7fabea6ac9b4.png)

And the unrolled component with k=4.

![screenshot from 2016-01-26 12-40-38](https://cloud.githubusercontent.com/assets/145425/12579701/2d9fbf0c-c42a-11e5-83b6-9be0dee7a29e.png)

**Comment by ekg** [comment]
Adding in the connections to the acyclic components:

![screenshot from 2016-01-26 17-27-52](https://cloud.githubusercontent.com/assets/145425/12586851/35b5cd8a-c452-11e5-93ec-2cc55e16fbb3.png)

For fun ad later reference, here's the input graph in json:

```js
{"node": [{"sequence": "A", "id": 1}, {"sequence": "B", "id": 2}, {"sequence": "C", "id": 3}, {"sequence": "D", "id": 4}, {"sequence": "E", "id": 5}], "edge": [{"from": 1, "to": 2}, {"from": 3, "to": 2}, {"from": 2, "to": 5}, {"from": 2, "to": 3}, {"from": 4, "to": 3}, {"from": 3, "to": 4}, {"from": 4, "to": 4}]}
```

I'm working now on carrying forward the node translation.

A big missing piece here is flipping inversions, which was part of the objective of unrolling. I have written a draft which does not handle them. There is already functionality in vg to do this, but I'm not sure it's appropriate. Perhaps we can unroll first and then unflip the reversing edges. I need to think about this.

**Comment by ekg** [comment]
Here is a sketch of the process to reduce the complexity of the trees that are generated by the unrolling process. We want to compress them into DAGs so as to minimize the alignment cost, which is proportional to the length of the labels of the graph.

Say we start with this tree, as obtained earlier:

![screenshot from 2016-01-26 19-27-34](https://cloud.githubusercontent.com/assets/145425/12590440/54b4a1f0-c463-11e5-814d-e2f699de5b85.png)

We then label all the nodes with their identity-relative height from the root. In other words, we label them with the number of times that we see a node with the same original identity before we reach them when walking from the root (here 1:B) This should explain:

![screenshot from 2016-01-26 12-40-38 labeled](https://cloud.githubusercontent.com/assets/145425/12590454/652e94f0-c463-11e5-8644-4d1327d0e3af.png)

The next step is to merge nodes that have the same original identity and identity-relative height. Because our objective is to reduce the complexity of the graph as much as possible, we merge the largest group of these first. In this case, we start with C<sub>2</sub>. We merge all of these into node 4:C, like so.

![screenshot from 2016-01-26 19-27-49](https://cloud.githubusercontent.com/assets/145425/12590535/c13dfa2e-c463-11e5-9aa6-6f106149857c.png)

Now, continuing, we re-label the graph with the identity-relative height. We would then merge the next-largest group, in this case 5:B, 9:B, and 16:B, which comprise the group B<sub>3</sub>, resulting in this graph:

![screenshot from 2016-01-26 19-28-03](https://cloud.githubusercontent.com/assets/145425/12590585/1070fc18-c464-11e5-913f-a157a064c582.png)

Now, we'd merge 6:D, 11:D, and 17:D, which comprise the group D<sub>4</sub>.

![screenshot from 2016-01-26 19-28-18](https://cloud.githubusercontent.com/assets/145425/12590604/24f78152-c464-11e5-9975-a01d31f7c51c.png)

Finally, we merge 10:C and 12:C, which form the group C<sub>3</sub>.

![screenshot from 2016-01-26 19-28-31](https://cloud.githubusercontent.com/assets/145425/12590632/45a84e22-c464-11e5-9b13-9ac2725fb566.png)

The resulting graph has no more groups of nodes at the same height with the same original identity, so we stop.

The starting graph is of length 20, while we finish with one of length 11. This is a substantial savings, and will be more substantial with larger k.

**Comment by ekg** [comment]
Here's an outline of the algorithm to determine the original identity rank of each node in the DAG simplification process.

We will store a vector R with length equal to the number of original nodes in the strongly connected component for each node we reach. For each node n in topological order: Compute R<sub>n</sub> as the element-wise max(R<sub>m</sub> for all node m in parents of n) or [ 0 0 0 ... ] where n has no parents and increment the vector element corresponding to the original identity of the node in the input graph.

At the end, we traverse all the nodes and label them with the value  the corresponding R<sub>n</sub> vector. This establishes the original identity ranks. We then use these as described in the previous post when collapsing the graph. In this way we avoid generating cycles and appropriately simplify the graph without modifying its sequence space.

**Comment by ekg** [comment]
I've coded up the tree simplification algorithm. Here goes for k=6 on the test graph we have here.

The size reduction is clear from looking at these. The first tree is length 65, the last is 18.

The algorithm does not seem very efficient to me. We could cache these unrolled regions. Computing them on the fly may be fast enough, but I don't have a sense of it.

![tree-2](https://cloud.githubusercontent.com/assets/145425/12647257/98ec5ecc-c5d4-11e5-8aff-280e6cd00f65.png)
![dag-2-0](https://cloud.githubusercontent.com/assets/145425/12647266/a2c62928-c5d4-11e5-845e-99c9a6b18d3a.png)
![dag-2-1](https://cloud.githubusercontent.com/assets/145425/12647268/a2c8a734-c5d4-11e5-9a4f-75ab682d6b0f.png)
![dag-2-2](https://cloud.githubusercontent.com/assets/145425/12647270/a2cf20e6-c5d4-11e5-8089-f1d0a188ac09.png)
![dag-2-3](https://cloud.githubusercontent.com/assets/145425/12647267/a2c89726-c5d4-11e5-8afe-f1aa3b92d28b.png)
![dag-2-4](https://cloud.githubusercontent.com/assets/145425/12647271/a2d5129e-c5d4-11e5-9af1-4cb11fdf69c9.png)
![dag-2-5](https://cloud.githubusercontent.com/assets/145425/12647269/a2cee338-c5d4-11e5-9003-84cfc56269b0.png)
![dag-2-6](https://cloud.githubusercontent.com/assets/145425/12647272/a2ddb4da-c5d4-11e5-95e9-a9a08577e4bf.png)
![dag-2-7](https://cloud.githubusercontent.com/assets/145425/12647273/a2e4d7d8-c5d4-11e5-862c-02ead00da664.png)
![dag-2-8](https://cloud.githubusercontent.com/assets/145425/12647274/a2e5948e-c5d4-11e5-9a39-383770d1b957.png)
![dag-2-9](https://cloud.githubusercontent.com/assets/145425/12647275/a2e9b73a-c5d4-11e5-951f-4e27edcfe8fb.png)
![dag-2-10](https://cloud.githubusercontent.com/assets/145425/12647276/a2ea6202-c5d4-11e5-8ce8-ea32a48c450a.png)

**Comment by ekg** [comment]
Outstanding issues.

1. Inversions should be handled correctly in the unrolling process by flipping the node when building up the initial tree and also recording the relative orientations in the translation. This will require a lot of small tweaks to what I've written in the unroll branch, but seems straightforward provided obvious extensions to the data structures that record the translation.

2. Where a strong component has multiple input nodes, we should check if the input nodes have partial order relative to each other and choose the top-ordered one when unrolling the component. Otherwise, we end up with extra redundancy in the graph. However, we can't do much if the entry points have no obvious order relative to each other.

3. Tests of unrolling should use kmer enumeration to verify that input and output graphs have the same sequence space. This is the objective and is easy to check.

**Comment by ekg** [comment]
I can now handle inversions, at least where these occur within strongly connected components.

Taking this graph:

![screenshot from 2016-02-04 17-03-09](https://cloud.githubusercontent.com/assets/145425/12822718/3a91730c-cb61-11e5-8bc9-6f8f5e101c12.png)

If we walk forward from B, we get to the reverse complement of C. Then we stay on the reverse strand until we get out by coming back from C reverse into B forward. It is the same as the previous test graph but the cyclic component is inverted.

Adding some special characters (instead of reverse complementing things) we can see the progression of the algorithm easily for k=4.

The original:

![screenshot from 2016-02-04 17-08-33](https://cloud.githubusercontent.com/assets/145425/12822984/2cf2a026-cb62-11e5-91ec-1c314a330820.png)

And the inverted:

![screenshot from 2016-02-04 17-09-17](https://cloud.githubusercontent.com/assets/145425/12822995/35471284-cb62-11e5-9c0a-46c1f21412b1.png)

I am going to test a few more cases of this, then push my commits to the pull request.

The next step will be to consider how to handle non-cyclic components with inversions. I don't know if the flipping process is handling these properly. The acyclic components could be handled using the same code so as to resolve inversions.

**Comment by ekg** [comment]
To test the process of linking the unrolled strongly connected component back into the graph, I modified the graph to include a loop that maintains the inversion, resulting in exit nodes that are inverted.

```
{"node": [{"id": 1, "sequence": "A"}, {"id": 2, "sequence": "B"}, {"id": 3, "sequence": "C"}, {"id": 4, "sequence": "D"}, {"id": 5, "sequence": "E"}], "edge": [{"from": 1, "to": 2}, {"from": 1, "to": 5}, {"from": 3, "to": 2, "from_start": true}, {"from": 2, "to": 5}, {"to_end": true, "from": 2, "to": 3}, {"to_end": true, "from": 3, "to": 2, "from_start": true}, {"from": 4, "to": 3}, {"from": 3, "to": 4}, {"from": 4, "to": 4}]}
```

![screenshot from 2016-02-04 22-30-07](https://cloud.githubusercontent.com/assets/145425/12831922/e1360c18-cb8e-11e5-8d2b-4b2445c166e1.png)

It appears that unrolling works on this and correctly links things back. (k=4)

![screenshot from 2016-02-04 22-28-27](https://cloud.githubusercontent.com/assets/145425/12831938/ecc6d6ca-cb8e-11e5-976d-f6c258f2eaed.png)

This is cool. The next step is to target inversions that lie outside of strongly connected components. I need to think about how to do this.

**Comment by adamnovak** [comment]
If it's not in a strongly connected component, then you can just flip any nodes you come across where you hit the end of the node when you were expecting the start, right? The ones outside strongly connected components seem like the easy ones.

**Comment by ekg** [comment]
@adamnovak let me try to explain what's confusing me.

I would like a function which maps an input graph to one that contains no cycles, and represents all sequences of length _k_ in the original graph on the forward strand. It may contain some sequences of length > _k_ that are found in the original graph, but not all those in strongly connected components.

To do this, I can first unroll strongly connected components with the approach described here so that the graph is acyclic and all sequences from strongly connected components up to length _k_ are represented. I have adapted this process to handle inversions and it seems stable. It does not handle inversions in the acyclic components of the graph.

**Approach 1:** We are representing the reverse complement of the graph implicitly. We could reverse complement the graph to obtain it. Inversions then become crossovers between the forward and reverse graph.

This will work, but it is costly. From the perspective of an aligner, aligning against the full reverse complement of the graph may duplicate effort. We may be certain of our orientation when we initiate the alignment. In vg we collect subgraphs in the same orientation as our read, so this will be the typical state of affairs. If I am reasonably sure of my orientation, I would still at least want to be sure that I can align up to k characters on the other side of the inversion.

**Approach 2:** I copy the reverse complement of the graph up to _k_ characters on the other side of an inversion into the graph and link it in so that it appears in the natural orientation of the graph.

This can do better than _k_ in some cases. If I have two subgraphs that are connected by a single inverting edge, there is no problem--- I can just choose the orientation of one graph as "forward", then flip the other and I generate a graph that shows me the inversion from the perspective of a sequence that passes through it.

However, if I have edges connecting to both the forward and reverse of the node in question, I can't do this because it introduces contradictions. The node will have to be duplicated as in approach 1 into a forward and reverse complement.

Here is what happens now.

There are no cycles in this graph:

![screenshot from 2016-02-06 13-27-10](https://cloud.githubusercontent.com/assets/145425/12866934/1dd2f4a8-ccd7-11e5-8a88-f08e360d4720.png)

(For clarity: There is an edge from 1:A to 9:B- that inverts, and then another from 9:B- to 5:E. It is easier to note than to try to get it to plot right.)

If I run this graph through `orient_nodes_forward`, I get a fragmented result:

![screenshot from 2016-02-06 13-43-14](https://cloud.githubusercontent.com/assets/145425/12866946/97d8ff0e-ccd7-11e5-98a8-a46613df7673.png)

These cases can't be handled by orient_nodes_forward, so it breaks the graph. We can still align against this graph, but the alignment is greedy and so we don't get a complete alignment, only the alignment across the best-matching subgraph.

**Comment by pgrosu** [comment]
Hi Erik,

Somehow I am confused about something - and this might be just me.  You mention:

> (For clarity: There is an edge from 1:A to 9:B- that inverts, and then another from 9:B- to 5:E. It is easier to note than to try to get it to plot right.)

But I see `9:C-` and `11:B-` in the top graph, with `9:C-` and `11:-B` in the bottom graph.  I am not seeing a `9:B-`.  Does the order of the minus sign make a difference?

Am I misreading something?

Thanks,
Paul

**Comment by pgrosu** [comment]
Hi Erik and Adam,

I was playing with this with some other code I put together, and was wondering about something.  If a node has both a forward and reverse direction, why not just use the same graph where for each alignment instance one of those edges is disabled (or temporarily removed).  For example, if `11:B-` has a direction coming from both `7:C-` and `1:A`.  Then you can run your alignment with the edge from `1:A` (with the one from `7:C-` disabled), and another alignment with the edge from `7:C-` (with the one from `1:A` disabled) on the same graph, without using any expensive space of multiple graphs.  This way the greedy alignment will work for the best-matching subgraph as intended on one connected graph, and should provide the optimal result.

Please let me know if I misinterpreted something.

Thanks,
Paul

**Comment by pgrosu** [comment]
In any case, below is the code I put together to test this, and hope it is okay if it is in Java.  Sorry if it looks a little hurriedly put together.  This file would be called `ekgDigraph.java`, in case you would like to run it:

```Java
import java.io.*;
import java.util.*;

public class ekgDigraph {
  
  Map<String, ArrayList<String>> adj = new HashMap<String, ArrayList<String>>();
  Set<String> dfsMarked = new HashSet<String>();
  int ConnectedComponentCount;
  Map<String, Integer> associatedComponent = new HashMap<String, Integer>();

  public void addEdge( String v, String w ) {
    ArrayList<String> currentAdjacentEdges = new ArrayList<String>();
    if ( adj.get(v) != null ) {
      currentAdjacentEdges = adj.get(v);
    } 
    currentAdjacentEdges.add(w);
    adj.put( v, currentAdjacentEdges );
  }

  public ArrayList<String> adj( String v ) {
    return adj.get(v);
  }


  public ekgDigraph( String fileName ) throws Exception {

    FileReader graphFile = new FileReader(new File(fileName));
    BufferedReader bufferedReader = new BufferedReader(graphFile);
    String oneEdge;

    int edgeCount = 0;
    int vertexCount = 0;
    String [] edge;
    while((oneEdge = bufferedReader.readLine()) != null) {

      edge = oneEdge.split(" ");
      if ( !edge[0].startsWith("#") ) {  // Skip edge starting with # sign, 
                                         // in order to have only a forward or reverse edge
                                         // into a node.

        int edge_count = 0;
        
        for( int idx = 0 ; idx < edge.length ; idx++ ) {
          if( edge[idx].trim().length() != 0 ) {
            edge[edge_count] = edge[idx].trim();
            edge_count++; // increment the count
          }
        }

        String v = edge[0];
        String w = edge[1];
        addEdge(v, w);
      }
    }

    bufferedReader.close();

  }

  public void findConnectedComponents() {

    ekgDigraph G = this;
    
    for ( String vertex : G.adj.keySet() ) {
      if ( !dfsMarked.contains( vertex ) ) {
        dfs( G, vertex );
        ConnectedComponentCount++;
      }
    }
  }

  public void dfs( ekgDigraph G, String vertex ) {

    dfsMarked.add( vertex );
    associatedComponent.put( vertex , ConnectedComponentCount);
    if ( adj.get(vertex) != null ) {
      for ( String adjVertex : adj.get(vertex) ) {
        if ( !dfsMarked.contains( adjVertex ) ) {
          dfs( G, adjVertex );
        } 
      }
    }
  }

  public int count() {
    return ConnectedComponentCount;
  }

  public void printComponents() {

    ekgDigraph G = this;

    System.out.println();
    System.out.println( "The graph has " + G.count() + " components.");
    System.out.println();

    ArrayList<String>[] components = (ArrayList<String>[]) new ArrayList[ G.count() ];

    for ( int i = 0 ; i < components.length ; i++ ) {
      components[i] = new ArrayList<String>();
    }

    for ( String vertex : associatedComponent.keySet() ) {
      components[ associatedComponent.get(vertex) ].add(vertex);
    }

    for ( int i = 0 ; i < components.length ; i++ ) {
      System.out.print( "Component " + i + " => " );
      for ( String v : components[i] ) {
        System.out.print( v +  " " );
      }
      System.out.println();
    }
  }

  public void joinComponents() {

    ekgDigraph G = this;

    ArrayList<String>[] components = (ArrayList<String>[]) new ArrayList[ G.count() ];

    for ( int i = 0 ; i < components.length ; i++ ) {
      components[i] = new ArrayList<String>();
    }

    for ( String vertex : associatedComponent.keySet() ) {
      components[ associatedComponent.get(vertex) ].add(vertex);
    }

    boolean [] isComponentConnected = new boolean[components.length];

    for ( int i = 0 ; i < components.length ; i++ ) {
      for ( String v : components[i] ) {
        if ( !isComponentConnected[i] && adj.get(v) != null ) {
          for ( String adjVertex : adj.get(v) ) {
            for ( int rest = 0 ; rest < components.length ; rest++ ) {
              if ( i != rest && components[rest].contains( adjVertex ) && !isComponentConnected[i] ) {
                System.out.println( "Connect component " + i + " with " + rest + " from node " + v + " -> " + adjVertex );
                isComponentConnected[i] = true;
                isComponentConnected[rest] = true;
                break;
              }
            }
          }
        }
      }
    }
  }

  public static void main( String [] args ) throws Exception {

    ekgDigraph G = new ekgDigraph( args[0] );
    
    G.findConnectedComponents();

    G.printComponents();

    // Print a separator line
    System.out.println();
    
    G.joinComponents();

  }

}
```

The graph file has the following contents, which I called `ekg-graph.txt`.  It is based on the above graph, with only edges `1A 11B` and `1A 26B` remarked out using `#`, so that I simulate turning off one edge on a node that has both a forward and a reverse direction:

```Bash
1A  6B
#1A 11B
1A 8B
1A 25A
#1A 26B
1A 5E
6B 5E
6B  7C
7C 8B
7C 11B
7C 23D
11B 12C
11B 5E
12C 13D
12C 8B
13D 17D
13D 14C
17D 14C
17D 15D
14C 15D
14C 8B
8B 9C
8B 5E
23D 28D
23D 9C
28D 9C
28D 27D
9C 27D
9C 25B
9C 26B
25B 5E
26B 5E
```
Below are the run results, which show a complete graph in that all the components connect to each other:

```Bash
> java ekgDigraph ekg-graph.txt

The graph has 5 components.

Component 0 => 14C 15D 8B 9C 5E 26B 27D 25B
Component 1 => 11B 12C 13D 17D
Component 2 => 23D 28D
Component 3 => 6B 7C
Component 4 => 1A 25A

Connect component 1 with 0 from node 11B -> 5E
Connect component 2 with 0 from node 23D -> 9C
Connect component 3 with 0 from node 6B -> 5E
Connect component 4 with 3 from node 1A -> 6B

>
```

Hope it helps,
~p

**Comment by ekg** [comment]
I'm calling the "flattening" operation "unfold", with the metaphor being that the inversions are the folds of the graph.

@pgrosu I think this is roughly the same thing that you implemented? I haven't had a chance to run your code.

I've written it up and have what looks like a stable prototype, modulo all the debugging code I've added and weird way that I'm handling the nodes and lengths on the graph (I'm counting 1 for each node as length and adding + and - to represent what operations are done).

Taking our friendly input graph:
 which resulted from unrolling:

![screenshot from 2016-02-04 22-30-07](https://cloud.githubusercontent.com/assets/145425/12831922/e1360c18-cb8e-11e5-8d2b-4b2445c166e1.png)

Which we've unrolled to length 3:

![screenshot from 2016-02-06 13-27-10](https://cloud.githubusercontent.com/assets/145425/12919847/d2e87684-cf3d-11e5-9857-9b56ddc2da0d.png)

We can unfold to length 50 (in this case, everything) and connect up with the accessible parts of the reverse complement of the graph. We then remove all inverting edges.

![screenshot from 2016-02-09 14-56-50](https://cloud.githubusercontent.com/assets/145425/12919874/ee25b5c4-cf3d-11e5-8ab7-5e83de850e3f.png)

The added `-` behind the node names means they've been flipped in this step.

So it looks to be working and I'm going to cleaning up the code and testing it more seriously. The unrolling algorithm needs to be improved to ensure we don't end up in runaway cases that generate huge initial trees prior to compaction.

**Comment by ekg** [comment]
I now need to test this as a setup for alignment. Then I will close the issue.

**Comment by pgrosu** [comment]
@ekg Awesome, glad to hear you got the problem solved :smile:  Basically what I was doing is letting the program decide the components, and performing [edge contraction](https://en.wikipedia.org/wiki/Edge_contraction) implicitly to get around the issue of unrolling outside a component.  Afterwards I was just stitching them back together, to ensure a complete graph.  Again, one can use those vertices that have both a forward and backward direction as a bifurcation without unrolling.  So whatever path approaches that node, then the alignment would be topological ordered by proceeding along the *1...n* edge(s) that continue in the same direction of the approaching edge.  The alignment would thus be recursive with a radius of k rounds in all satisfiable directions without the need to unroll.

**Comment by ekg** [comment]
@adamnovak so you were right, and the pathological cases occurred very quickly. I'm stubborn though so I followed through up to the point where things break. I have some hacks that prevent unrolling when there is too much branching (as branching makes the backtracking search blow up) but this isn't ideal and I'm following suggestions from @richarddurbin.

Basically, we unroll the graph "vertically" by copying it and connecting nodes in the first copy to nodes in the second copy where the original graph had such an edge. We then continue this process from the "newest" nodes. The result is a DAG, and it may be possible to run a phase of compression to remove redundancy and reduce the alignment cost. I'll detail the process here as soon as I've got the code to demonstrate them visually. The problem is that it may be expensive to determine when we've unrolled far enough to support queries of a given length. This could be tackled using dynamic programming to find the shortest length path through the unrolled graph, but I haven't worked this out yet.

Here's to being stubborn and trying the impossible :grin: The good news is a lot of the code that I wrote in support of the backtracking solution that @jltsiren and I worked out will be applicable to the O(NM) unroll.

**Comment by pgrosu** [comment]
@ekg Is it an issue of keeping it in memory or storing it, and afterwards performing the joins?  How big is the tree - 100s of GBs, several terabytes, petabytes? Those workloads can be easily offset with the tree being indexed on disk.  I'm not seeing why this bottleneck cannot be easily solved.  Sure the shortest path algorithms (i.e. Dijkstra's, A* search, etc.) are fairly common approaches.

Thanks,
~p

**Comment by ekg** [comment]
@pgrosu The trees grow exponentially with the number of branching edges in the graph, and could quickly approach scales that are simply unthinkable. @jltsiren expressed concern about this and I went ahead and tried it anyway to see what would happen on real data. In real data the trees get very big very quickly. So I am going back and doing the sensible thing with an algorithm that will increase the time and space costs of aligning against strongly connected components only linearly with the length of sequence we want to align against them.

**Comment by pgrosu** [comment]
@ekg What would be your longest aligning sequence?  Yes, I know that trees can grow exponentially :)  If very large, why not generate the tree and unroll locally for small sections of the alignment, and the iterate only on paths of subtrees that show stronger alignment and then ignore the other unrolls.  What I am trying to say, is that some paths on the tree are more important than others, and the initial alignment could dictate which paths to keep unrolling on - basically alignments are not random.

~p

**Comment by richarddurbin** [comment]
I still think there is a much more efficient way to do this dynamically on the graph.
Outer loop along the sequence to be aligned.  Inner loop across possible nodes in the subgraph.  The complication is deletions from the graph,
i.e. the sequence is missing something in the graph.  One way to handle this is to loop multiple times over the inner loop per outer loop, each
addititional one being a single base deletion.  It is possible to make this more efficient by creating new subgraphs that skip multiple steps - I
think there are several ways to do this.

Richard

> On 15 Feb 2016, at 05:21, Erik Garrison <notifications@github.com> wrote:
> 
> @pgrosu <https://github.com/pgrosu> The trees grow exponentially with the number of branching edges in the graph, and could quickly approach scales that are simply unthinkable. @jltsiren <https://github.com/jltsiren> expressed concern about this and I went ahead and tried it anyway to see what would happen on real data. In real data the trees get very big very quickly. So I am going back and doing the sensible thing with an algorithm that will increase the time and space costs of aligning against strongly connected components only linearly with the length of sequence we want to align against them.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/ekg/vg/issues/182#issuecomment-184068777>.
> 




-- 
 The Wellcome Trust Sanger Institute is operated by Genome Research 
 Limited, a charity registered in England with number 1021457 and a 
 company registered in England with number 2742969, whose registered 
 office is 215 Euston Road, London, NW1 2BE.

**Comment by pgrosu** [comment]
@richarddurbin Or one can perform it the other way around for deletions, by taking paths along the graph and aligning them to the sequence.  Then the gap(s) will show up naturally :)

~p

**Comment by ekg** [comment]
@richarddurbin 

> I think there are several ways to do this.

There definitely are.

I am doing the DAGification up front in such a way as to maintain the sequence space of the original graph in the DAGified version up to a certain length. If this length is >= the length of our read, then we can be guaranteed the correct alignment against the first graph.

I think this approach makes sense, because I'd like to maintaining compatibility with POA, for which there are a number of SIMD-based implementations that provide large improvements in performance over non-SIMD implementations. [GSSW](github.com/ekg/gssw) generalizes Farrar's striped smith waterman to work on graphs. At the hackathon in Sendi I've met Hajime Suzuki, who has (as I understand from our conversations) has generalized a variant of [Myers' bit-vector algorithm for approximate string matching](http://dl.acm.org/citation.cfm?id=316550) to handle gaps. He is currently working on a manuscript describing that and [further extending his method to work on graphs](https://bitbucket.org/suzukihajime/aln2) (see branch v3). I would like to be able to leverage his work once it is ready, and doing so would be simplest to implement via DAGification of the graph.

There are other ways to do it, and it seems to me there may be some regimes in which they will be preferable, but for the time being I am interested in minimizing effort and using available components as much as possible.

I have resolved alignment against cyclic and inverting graphs in the following way. I will illustrate with the example graph from before. In the test directory, `vg view -d graphs/atgc.vg`:

![screenshot from 2016-02-04 22-30-07](https://cloud.githubusercontent.com/assets/145425/12831922/e1360c18-cb8e-11e5-8d2b-4b2445c166e1.png)

We rebuild the graph, first copying over all acyclic components of the graph, and then adding in a dagified version of each strongly connected component.

For each strongly connected component (SCC), we make a copy of the nodes without any edges. To "unroll", we add a new copy of the SCC, and for each edge in the SCC, we add an edge from the node in the first copy to that in the second. At each step, we track the minimum distance from a given node to a node in the first copy of the SCC. When the minimum of these distances is greater than the length of the sequence we want to query against the graph, we stop. Alternatively, we may unroll as many steps as desired.

The first step:

![screenshot from 2016-02-23 16-19-48](https://cloud.githubusercontent.com/assets/145425/13244430/914790a4-da49-11e5-808a-b39a0f64a193.png)

After two steps:

![screenshot from 2016-02-23 16-20-44](https://cloud.githubusercontent.com/assets/145425/13244447/9c1ee8ba-da49-11e5-98c9-4dccde97eb32.png)

After 3 steps:

![screenshot from 2016-02-23 16-21-13](https://cloud.githubusercontent.com/assets/145425/13244453/a4c31450-da49-11e5-9649-5bd5f011f82f.png)

After 4 steps:

![screenshot from 2016-02-23 16-21-32](https://cloud.githubusercontent.com/assets/145425/13244459/b11ed9f0-da49-11e5-92d3-01b55c6f3ab0.png)

If we first apply the unfold process to the graph, we can ensure that there are no inverting edges and the reverse complement of the graph is available to reached out to our query length _k_. Then this dagify unrolling process will not yield any incidental cycles and will be generate a graph in which a read of length <= _k_ can be aligned to the forward complement of the graph.

Note that we maintain a node translation so that alignments against this graph can be projected back into the source graph. The translation of these is trivial.

In summary, this procedure provides a linear time unrolling process, which does not blow up in space and time. I've implemented this and pulled it into master, and am testing it on the MHC.

**Comment by vnsk** [comment]
Can anyone please help me with the vg commands to visualize the DAG of local vg which is fed as an input to gssw module ? There are so many flags associated with vg mod command which deals with one step at a time for unfolding, breaking..etc. What command should be used to visualize the complete DAG of a local vg which is ready to be given as an input to gssw algorithm ?

**Comment by adamnovak** [comment]
I think it would be something like `vg mod --unfold <length of read> --dagify-to <length of read>`. If you really want to see the graph, I think there might be code to dump the actual graph being fed into gssw, behind an `#ifdef debug` that you could remove.

**Comment by vnsk** [comment]
Thanks. Also can you tell how to run gssw_aligner.cpp module on a separate DAG which is outputted by   ` vg mod --unfold <length of read> --dagify-to <length of read>` with a separate read sequence ?

**Comment by vnsk** [comment]
Also, I am trying to load a subgraph using the vg find command. I am using like this
`  ./vg find -r 1:6 -x tiny.xg | ./vg view -d - > viz.dot` 

But the edges are not shown in the this subgraph. How can I visualize the nodes and edges with orientations of a subgraph ?

**Comment by adamnovak** [comment]
I don't think we have code to let you dump an unmodified graph directly
into GSSW.

To get some edges out of "vg find" I think you need to set a context number
of hops to go out from the selected nodes. Try adding `-c 1` to the `find`
command.

**Comment by vnsk** [comment]
Hi, I tried adding `-c 1` but still the complete sub-graph(containing nodes and edges) is not extracted. Please let me know how can we extract a complete sub graph (containing both nodes and edges ) between either range of two nodes or by some other method. 
 
Also, in the above query, I understand we cannot dump unmodified graph directly into gssw. My question is how to dump the DAG output of any local vg using the 
`vg mod --unfold <length of read> --dagify-to <length of read>`
and run the gssw_aligner.cpp module on it independently and check the result of gssw_align on it.

**Comment by adamnovak** [comment]
Hm, I think `-c 1` is supposed to not leave any dangling edges. Can you
open another bug for that issue?

As for how to get the output of `vg mod` into `gssw_aligner.cpp`, we
currently have no command line interface to do it. You would need to write
some C++ code that reads in a graph (probably just as a plain Protobuf
`vg::Graph` and not as a fully cross-indexed `vg::VG`), creates a
`vg::Aligner`, and runs the `align()` method on the aligner to align
something against the graph. If you want to write a new subcommand that
does that, you would put your main cpp file in the `subcommands` directory,
make sure it registers itself with the subcommand system like the other
subcommands in there, and then modify the Makefile to build it and link the
resulting object file into the main vg binary.

If you just want to add your own binary to the build that just holds your
C++ code, you would link against `libvg.a`, which the build process dumps
all the important non-CLI stuff into.

If you are encountering a specific problem, it might make the most sense to
add a C++ unit test under the unittests directory, which (if you put it in
an existing file or set up a new file right in the Makefile) would get
linked into vg and be runnable with `vg test`.

**Comment by vnsk** [comment]
Okay. Thanks @adamnovak ! I have one more question regarding the GFA file format of vg. The Path line 'P' is different from any other standard GFA file formats. Could you please explain the following Path line "P' in the context of vg GFA file. Here is the sample line from tiny.gfa file of vg 
`P       6       x       4       +       3M` How to interpret this?

**Comment by adamnovak** [comment]
I think @ekg might have to explain that one. I know that each P line
represents a visit of a path to a node. The "x" is the path name, the "+"
is the orientation in which the path visits the node, and the "3M" is a
CIGAR giving the alignment between the path and the node. Of the "6" and
"4", one is node ID being visited, and one is the index of the visit along
the path, but I don't remember which is which.

It's not really proper GFA; the mainline GFA people decided to go with
single-line representations of paths, which don't necessarily work for our
use case where a path may visit millions of nodes.

**Comment by vnsk** [comment]
@adamnovak I have one doubt regarding GFA representation of variation graph. In "Link line (L line)" of GFA  if there is path `A +  B -`  , how should we interpret it ?

1) According to standard GFA file, this L line is explained as path from A forward strand to reverse complement of B 

2) While according to VG format understanding of mine it is path from B forward strand to A forward strand  as well as path from A reverse complement to B reverse complement since VG graph is bidirectional.  

Can you please clarify as to which interpretation is correct  ?

**Comment by edawson** [comment]
Hey @vnsk ,
   I'm going to try to tackle your GFA questions. You're correct that the GFA path syntax has moved to one path per `P` line, with all path objects list on that line. We're frozen on a version before they implemented this change while we wait for the syntax to firm up.

> Here is the sample line from tiny.gfa file of vg
P 6 x 4 + 3M How to interpret this?

We use `P` as an annotation on a node,where for each segment on the path there is a corresponding `P` line. This is in contrast to GFA's treatment of paths as first-class objects. So the above example would translate to "Node 6 is on path x; it is the fourth item (rank) in the path; it is in the forward direction; all three bases in 6 are on the path." In some cases, A single path line `P` in GFA syntax could be generated by collapsing all the vg `P` lines to one and modifying the cigar as needed, and I plan to support this in the future.

The reason we did this is that our graphs (as @adamnovak said) may have millions of items in a path, and cramming them all on one line makes the GFA format completely illegible to human beings. We've submitted [a pull request](https://github.com/GFA-spec/GFA-spec/pull/47) to implement a `W` or `walk` line, but this has not yet been approved.

> I have one doubt regarding GFA representation of variation graph. In "Link line (L line)" of GFA if there is path A + B - , how should we interpret it ?


This is actually in the style of proper GFA (assuming I got the flags right). VG does allow directionality in graphs (though it doesn't require it), so in this case you're correct in interpreting it as:

> this L line is explained as path from A forward strand to reverse complement of B

i.e. an edge from the "tail" of node A to the "tail" of node B.

Hope this helps clarify things!

**Comment by ekg** [comment]
By the way, the vg align command does nearly what's wanted with respect to
feeding the graph directly into gssw. We could try to make the input more
"raw" but I am a little confused as to why this is needed. If the graph is
acyclic then it is passed into gssw unmodified.

**Comment by vnsk** [comment]
Hi @ekg, @adamnovak, @edawson 
..I have on doubt regarding the GFA conversion of VG project. I considered the following sample fasta and vcf file. 

```
sample.fasta 
>x
CAAATAAGGCTTGGAAATTTTCTGGAGTTCTATTATATTCCAACTCTCTGCAAATAAGGCTTGGAAATTTTCTGGAGTTCTATTATATTCCAACTCTCTG
```
And the VCF file as follows 
```
sample.vcf file 
##fileformat=VCFv4.1
##fileDate=20141110
##source=mutatrix population genome simulator
##seed=1415643582
##reference=x.fa
##phasing=true
##commandline=mutatrix --dry-run -s 0.05 -i 0.01 -p 2 x.fa
##INFO=<ID=AC,Number=A,Type=Integer,Description="Alternate allele count">
##INFO=<ID=TYPE,Number=A,Type=String,Description="Type of each allele (snp, ins, del, mnp, complex)">
##INFO=<ID=SVTYPE,Number=A,Type=String,Description="Type of SV">
##INFO=<ID=NS,Number=1,Type=Integer,Description="Number of samples at the site">
##INFO=<ID=NA,Number=1,Type=Integer,Description="Number of alternate alleles">
##INFO=<ID=LEN,Number=A,Type=Integer,Description="Length of each alternate allele">
##INFO=<ID=SVLEN,Number=A,Type=Integer,Description="Length of SV">
##INFO=<ID=MICROSAT,Number=0,Type=Flag,Description="Generated at a sequence repeat loci">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	X
x	3	.	A	<DUP>	99	.	SVTYPE=DUP;SVLEN=10	GT	1|1
x	5	.	T	<INV>	99	.	SVTYPE=INV;SVLEN=2	GT	1|1
x	9	.	G	<DEL>	99	.	AC=1;LEN=1;NA=1;NS=1;SVTYPE=DEL;SVLEN=2	GT	1|0
x	20	.	T	<DEL>	99	.	AC=1;LEN=1;NA=1;NS=1;SVTYPE=DEL;SVLEN=7	GT	1|0
x	29	.	T	C,G	99	.	AC=1;LEN=1;NA=1;NS=1;	GT	1|0
x	32	1	A	ATTTTAAAA	99	.	SVTYPE=INS;SVLEN=0	GT	1|1
x	35	.	A	C	99	.	AC=1;LEN=1;NA=1;NS=1;	GT	1|0
```
To construct and view corresponding GFA file using vg project. I Did the following:
```
./vg construct -r sample.fa -v sample.vcf.gz > sample.vg 

./vg view -g sample.vg > sample.gfa 

```
The output sample.gfa file is as follows:
```

sample.gfa 
H	VN:Z:1.0
S	1	CAAATAAGGCTTGGAAATTTTCTGGAGT
P	1	x	1	+	28M
L	1	+	2	+	0M
L	1	+	3	+	0M
L	1	+	4	+	0M
S	2	T
P	2	x	2	+	1M
L	2	+	5	+	0M
S	3	C
L	3	+	5	+	0M
S	4	G
L	4	+	5	+	0M
S	5	CTA
P	5	x	3	+	3M
L	5	+	6	+	0M
L	5	+	7	+	0M
S	6	TTTTAAAA
L	6	+	7	+	0M
S	7	TT
P	7	x	4	+	2M
L	7	+	8	+	0M
L	7	+	9	+	0M
S	8	A
P	8	x	5	+	1M
L	8	+	10	+	0M
S	9	C
L	9	+	10	+	0M
S	10	TATTCCAACTCTCTGCAAATAAGGCTTGGAAATTTTCTGGAGTTCTATTATATTCCAACTCTCTG
P	10	x	6	+	65M
```
Could anyone please help why the output GFA file failed to capture the first four lines of VCF file of <DUP>,<INV>, <DEL>, <DEL> svtypes ?

**Comment by adamnovak** [comment]
That's because vg construct doesn't actually know how to read symbolic
alleles like <DUP>, <INV>, or <DEL> yet (at least by default). @edawson has
been working on it.

I thought I was going to need this functionality too, but I managed to get
someone to articulate the VCFs I need using fully specified allele base
strings.

**Comment by edawson** [comment]
VG can parse <DEL> and <INV> tags if you pass the `-S` flag to construct and your VCF provides the `SVLEN` and `SVTYPE` in the INFO columns. It will parse <INS> if you pass `-S` and `-I insertions.fasta`, where insertions.fasta is a fasta file containing the insertion sequence of your VCF variants, and the tag in the alt field is the insertion id in "<>" brackets.

Check out the first part of this wiki page: https://github.com/vgteam/vg/wiki/SV-calling-with-vg

Please let me know if you hit any issues!

**Comment by adamnovak** [comment]
Should we consider making -S the default?

**Comment by edawson** [comment]
Probably yes. I've avoided it so far because it invalidates almost every test and has a few limitations (requires flat alternates and that the variant be at a position >1bp beyond the sequence start, for example).

**Comment by vnsk** [comment]
Hi @edawson ,
I tried to generate GFA using the command you suggested.

USING VG: 
```
./vg construct -f -S -a -r sample.fa -v sample.vcf.gz > sample.vg 
./vg view -g sample.vg 
```
1. The output gfa file is not capturing DUP (Duplicates). It seems that VCF line at POS 3 is ignored in the output gfa file.
2. Also INV (inversion) is not captured according to the edge convention. Since '-' edge sign was not there. Without the '-' sign the inversion cannot be understood from the GFA file ? Correct me if I am wrong . 

Following is the output GFA from vg for sample.fa and vcf shared above in the discussion thread.  
```
sample.gfa 
H	VN:Z:1.0
S	1	CAAA
P	1	x	1	+	4M
L	1	+	2	+	0M
L	1	+	3	+	0M
S	2	AT
P	2	_alt_ae86adfba5b06f8b4ddc1cda4a4650aedc380635_1	14	+	2M
L	2	+	4	+	0M
S	3	TA
P	3	_alt_ae86adfba5b06f8b4ddc1cda4a4650aedc380635_0	13	+	2M
P	3	x	2	+	2M
L	3	+	4	+	0M
S	4	AG
P	4	x	3	+	2M
L	4	+	5	+	0M
L	4	+	6	+	0M
S	5	G
P	5	_alt_9415f505f043a840c914dd0ab2bb0a5fcd8ec1b1_1	16	+	1M
L	5	+	7	+	0M
S	6	GCT
P	6	_alt_9415f505f043a840c914dd0ab2bb0a5fcd8ec1b1_0	15	+	3M
P	6	x	4	+	3M
L	6	+	7	+	0M
S	7	TGGAAATT
P	7	x	5	+	8M
L	7	+	8	+	0M
L	7	+	9	+	0M
S	8	T
P	8	_alt_e7ebb7fe1a868b189725dbd95eada2ac65e747f2_1	18	+	1M
L	8	+	10	+	0M
S	9	TTCTGGAG
P	9	_alt_e7ebb7fe1a868b189725dbd95eada2ac65e747f2_0	17	+	8M
P	9	x	6	+	8M
L	9	+	10	+	0M
S	10	T
P	10	x	7	+	1M
L	10	+	11	+	0M
L	10	+	12	+	0M
L	10	+	13	+	0M
S	11	C
P	11	_alt_3a321436cc62676f7a1d8ec28395c00ebe646587_1	20	+	1M
L	11	+	14	+	0M
S	12	G
P	12	_alt_3a321436cc62676f7a1d8ec28395c00ebe646587_2	21	+	1M
L	12	+	14	+	0M
S	13	T
P	13	_alt_3a321436cc62676f7a1d8ec28395c00ebe646587_0	19	+	1M
P	13	x	8	+	1M
L	13	+	14	+	0M
S	14	CT
P	14	x	9	+	2M
L	14	+	15	+	0M
L	14	+	16	+	0M
S	15	ATTTTAAAA
P	15	_alt_6337c431f778e028e6c96110e02be12547f8fa0b_1	23	+	9M
L	15	+	17	+	0M
S	16	A
P	16	_alt_6337c431f778e028e6c96110e02be12547f8fa0b_0	22	+	1M
P	16	x	10	+	1M
L	16	+	17	+	0M
S	17	TT
P	17	x	11	+	2M
L	17	+	18	+	0M
L	17	+	19	+	0M
S	18	C
P	18	_alt_705880aaca0ac6461baf914b1166a4023ebccd36_1	25	+	1M
L	18	+	20	+	0M
S	19	A
P	19	_alt_705880aaca0ac6461baf914b1166a4023ebccd36_0	24	+	1M
P	19	x	12	+	1M
L	19	+	20	+	0M
S	20	TATTCCAACTCTCTGCAAATAAGGCTTGGAAATTTTCTGGAGTTCTATTATATTCCAACTCTCTG
P	20	x	26	+	65M
```
These two warning messages were generated while using the construct command 
```
warning:[vg::Constructor] Unsupported variant allele "<CN0>"; Skipping variant(s)!
warning:[vg::Constructor] Unsupported variant allele "<DUP:TANDEM>"; Skipping variant(s)!
```
Also I tried generating GFA file for some other examples apart from the one mentioned. But as we could see this example (sample.vcf) is not actually containing any of warned alle.

**Comment by edawson** [comment]
Hey @vnsk ,

Thanks for trying this out. I haven't added handling for DUP type tags yet, because I haven't figured out how I want them to go in the graph (i.e. as cycles, or as unrolled repeats). I'm open to suggestions and could also just implement something if you want to give it a try.

INV tags are captured, and should be in the graph. Instead of using reversing edges to represent INV I'm duplicating the ref sequence, reversing it, then adding it as a node at the relevant position. Here's the source code:

          this->ref = fasta_reference.getSubSequence(this->sequenceName, this->position, sv_len);
          string alt_str(fasta_reference.getSubSequence(this->sequenceName, this->position, sv_len));
          reverse(alt_str.begin(), alt_str.end());
          this->alt[alt_pos] = alt_str;

If we look at this graph using dot (save GFA to x.gfa; `vg view -Fd x.gfa | dot -T pdf -o o.pdf`) , it looks like nodes [2,3] define your inversion:

[o.pdf](https://github.com/vgteam/vg/files/773858/o.pdf)

**Comment by vnsk** [comment]
Hi @edawson 

It would be great if DUP can be represented as cycles. Since `vg mod` command already converts duplications into unrolled repeats to a desired length of 'k', to better interpret the variants, it's beneficial to visualize DUP as cycles in the graph instead of unrolled repeats.

**Comment by vnsk** [comment]
Hi @edawson, @ekg , @adamnovak 

Is there any way to visualize the Short Tandem Repeats (Variation Type = STR) as loops using vg commads. 

I considered `chromosome1.fasta` file and variant calls of STR on chromosome1 from `sgdp.calls.vcf.gz`.  I tried the following command to  visualize the corresponding variation graph.

`./vg construct -f -S -a -r chromosome1.fasta  -v sgdp.calls.chr1.vcf.gz > graph.vg`

Surprisingly, the output graph is a complete Directed Acyclic Graph (DAG (without any cycles)) which should not be the case if tandem repeats are involved naturally. 

But how can one visualize the Short Tandem Repeats (STRs) as effectively as loops or cycles using vg?

**Comment by adamnovak** [comment]
Are these variants with an "STR" as the "SVTYPE"? Can you maybe post one (on a new issue?).

I think if you want it to work right now you might have to go out and write the "STR" case handling yourself in the code we use to parse variants. The code for construct is [here in vcflib](https://github.com/vcflib/vcflib/blob/30e61049efdbc1511a6500ca62c59378cb781467/src/Variant.cpp#L170-L175).

It might not quite fit the model we're using, because that code is in our VCF parser, vcflib, which just treats each variant as a set of aligned string alternatives, and doesn't really support cycles. So to make it really work properly you might have to add some special-case code in vg construct, to interpret theses as loops and not just plain alternation between specific numbers of copies.

The other alternative is to not use vg construct, and instead write your own code to generate a JSON or GFA-format graph describing your repeat, and then
